segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,204 @@
+     * Notes:
+     * * Only changes something if throughput changed significantly since the last time.
+     * * Changes the degree of parallelism in the same direction as throughput.
+     * ** Exception is if both throughput and the fill ratio are declining, in which case nothing is done.
+     *
+     * @return The degree of parallelism, whether it was updated or not.
+     */
+    int getOrUpdateParallelism() {
+        // Calculate the most recent throughput and Fill Ratio (of those items that have just been written).
+        final long time = this.timeSupplier.get();
+        final long elapsedMillis = (time - this.lastSnapshot.get().timeStamp) / AbstractTimer.NANOS_TO_MILLIS;
+        if (elapsedMillis >= STALE_MILLIS) {
+            // Too long since the last update; the data we have is no longer relevant. Reset.
+            resetSnapshot(time);
+        } else if (elapsedMillis >= UPDATE_FREQUENCY_MILLIS) {
+            // Enough time elapsed since the last update; update the snapshot.
+            updateSnapshot(time, elapsedMillis);
+        }
+
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the current snapshot based on the current data.
+     *
+     * @param time          The current time, in millis.
+     * @param elapsedMillis The elapsed time, in millis, since the last time this method ran.
+     */
+    private void updateSnapshot(long time, long elapsedMillis) {
+        final int recentCount;
+        final long recentLength;
+        synchronized (this) {
+            recentCount = this.recentWriteCount;
+            recentLength = this.recentTotalWrittenLength;
+            this.recentWriteCount = 0;
+            this.recentTotalWrittenLength = 0;
+        }
+
+        final double recentThroughput = (double) recentLength / elapsedMillis;
+        final double recentFillRatio = WriteQueue.calculateFillRatio(recentLength, recentCount);
+        final Snapshot lastSnapshot = this.lastSnapshot.get();
+        final double throughputDifference = recentThroughput - lastSnapshot.throughput;
+
+        int parallelism = lastSnapshot.parallelism;
+        if (Math.abs(throughputDifference / lastSnapshot.throughput) >= SIGNIFICANT_DIFFERENCE) {
+            // Throughput changed significantly.
+            // But first, check for a special case. If both Throughput and Fill Ratio are decreasing, then most likely the
+            // former is a result of the latter. As such, don't do anything, in that case.
+            boolean fillRatioDecreased = (recentFillRatio - lastSnapshot.fillRatio) <= -SIGNIFICANT_DIFFERENCE;
+            if (throughputDifference > 0 || !fillRatioDecreased) {
+                // Either the throughput increased or the Fill Ratio did not decrease significantly.
+                // Change the parallelism in the same direction as Throughput.
+                // Update the degree of parallelism, but make sure we don't exceed the given bounds.
```

Comment:
Suppose the optimal parallelism is 4, the min is 3 and the max is 10.
It starts at 6 and is increased to 7 because it was idle too long.
It then decreased the throughput. so it is decreased back to 6.
This increases the throughput, so it increases the parallelism back to 7. 
The parallelism will oscillate between 6 and 7 never finding the optimal value of 4.

Reply:
The throughput, once it reaches the "optimal value", plateaus and doesn't increase anymore. There's some code below which checks to see if this has stayed in the same place for too long, and if so, it will nudge it in some direction to try to get it out of such "local optimums".

Reply:
True, but my point is it moves in the wrong direction. The way the algorithm works, it assumes that if it decreases the parallelism by one, it is guarenteed to be below the optimal value. Hence it will assume that any increase in throughput resulting means that more should be sent in parallel. When it is equally likely that decreasing the number being sent in parallel is the CAUSE of the increase in throughput because the number that is being used is simply too high. In such a case there is no way for the throughput to ever decrease by more than 1. Because if ever there is a interval that has higher throughput, even if this is a direct result of having fewer parallel requests, it will increase the number of parrallel requests.

Reply:
Keeping everything else constant, throughput as a function of parallelism will increase up to a point, after which it will plateau and not increase or decrease. At that point, the throughput stays constant because we saturated the connection, and latency starts to increase. 

As such, decreasing the parallelism can either have no effect (if we are way above the optimal one) or decrease the throughput itself. 

Once it reaches a below-optimal parallelism, throughput will also be lower than the "optimum", which will cause more operations to queue up upstream, and hence have larger writes. Larger writes, keeping all else constant, will lead to an increase in throughput which will kick up the parallelism back to its optimal value. 

In order to fully understand the heuristic here, you need to consider how the OperationProcessor works too. Its goal is to lower latency when it can, and optimize throughput when it has no other choice. As such, if the BK queue size is less that the parallelism it will send the new operations down to BK as soon as they're received, which leads to smaller writes (i.e., smaller Fill Ratios). If the BK queue size exceeds the degree of parallelism, it will artificially delay its own processing by a few milliseconds in order to gather more operations and generate larger writes (i.e., larger fill ratios).